blocks/model.py

```diff
@@ -1,75 +1,105 @@
+:class:`Model` extends :class:`blocks.graph.ComputationGraph` :class:,
+which is able to handle annotations and roles in general, but is
+deliberately made unaware of specific annotations that a Theano graph
+created by Blocks typically has, such as bricks and application calls.  The
+:class:`Model` adds this functionality. Using :class:`Model` you can do
+things like quering all the bricks used to build the computation graph,
+requesting "hierarhical names" of the parameters, etc.
+
+For more information, see :class:`Model` docstring.
 
 """
 import logging
-from abc import ABCMeta, abstractmethod
 from collections import OrderedDict, Counter
 from itertools import chain
 
-from six import add_metaclass
-
 from blocks.graph import ComputationGraph
 from blocks.select import Selector
 from blocks.filter import get_brick
 
 logger = logging.getLogger(__name__)
 
-multiple_message = """
-
-Model with multiple outputs are currently only partially supported \
-in Blocks. For instance a call of 'get_objective' will crash. \
-Contact Blocks developers for more details.
-"""
-
-
-@add_metaclass(ABCMeta)
-class AbstractModel(object):
-    """A parameterized entity trained in the main loop.
 
-    A model is a parameterized entity the user trains a in a main loop.
-    The following are traits of every model:
-
-    * It has parameters and supports a way to access them. In addition
-      to returning handles to parameter objects it can return their values
-      as numpy arrays and set their values to given numpy arrays.
-
-    * It has an optimality objective.
-
-    * It can be serialized and deserialized by mean of pickling.
-
-    * It might have bricks as its components.
-
-    This class provides an interface for models. For experiments use
-    a subclass, e.g. the :class:`Model`.
+class Model(ComputationGraph):
+    """Handles annotations in Blocks-built computation graphs.
+
+    Use this class to handle your Blocks-created computation graph.
+
+    Examples
+    --------
+    >>> from theano import tensor
+    >>> from blocks.bricks import MLP, Tanh
+    >>> x = tensor.matrix('x')
+    >>> mlp = MLP([Tanh(), Tanh()], [10, 10, 10])
+    >>> y = mlp.apply(x)
+    >>> model = Model(y)
+    >>> # With model you can get access to the brick hierarchy
+    >>> model.get_top_bricks() #doctest: +ELLIPSIS
+    [<blocks.bricks.MLP object at ...]
+    >>> # With model you get "hierarchical" names for the parameters,
+    >>> # which encode the position of the owning brick in the
+    >>> # brick hierarchy.
+    >>> model.get_parameter_dict() #doctest: +NORMALIZE_WHITESPACE
+    OrderedDict([('/mlp/linear_1.b', b), ('/mlp/linear_0.b', b),
+    ('/mlp/linear_0.W', W), ('/mlp/linear_1.W', W)])
 
     """
-    @abstractmethod
+    def __init__(self, *args, **kwargs):
+        bricks = [get_brick(var) for var
+                  in self.variables + self.scan_variables if get_brick(var)]
+        children = set(chain(*(brick.children for brick in bricks)))
```

Comment:
Huh, didn't know you could pass an iterator as *args. Does this work on Python 2?